### PR TITLE
UI Automation in Windows Console: improve reliability of visible range checks

### LINKED ISSUE #9891: UI Automation in Windows Console: text review behaves inconsistently when the window is maximized
### Steps to reproduce:
1. Open Command Prompt with UIA enabled.
2. Maximize the window.
3. In a git repository, run `git log`.

### Actual behavior:
Text review gets stuck.

### Expected behavior:
Text review shows all the visible text.

#### NVDA installed/portable/running from source:
Source and installed.

#### NVDA version:
Latest master.

#### Windows version:
Windows 10 1903.

#### Name and version of other software in use when reproducing the issue:
Command Prompt/Windows Console

#### Does the issue still occur after restarting your PC?
Yes.

#### Have you tried any other versions of NVDA? If so, please report their behaviors.
No.

### PR DESCRIPTION
**Pull request by codeofdusk**
### Link to issue number:
Builds on #9614. Supersedes #9735 and #9899. Closes #9891.

### Summary of the issue:
Currently:

* When the console window is maximized (or the review cursor is otherwise placed outside the visible text), text review is not functional.
* The review top line and review bottom line scripts do not function as intended.

### Description of how this pull request fixes the issue:
This PR backports some functionality from #9735. In particular:

* The `isOffscreen` property has been implemented as `UIAUtils.isTextRangeOffscreen`.
* When checking if the text range is out of bounds, we now also check that oldRange is in bounds before stopping movement.
* Re-implemented `POSITION_FIRST` and `POSITION_LAST` in terms of visible ranges.

### Testing performed:
Repeated testing from #9899 and verified that the review top/bottom scripts are functional.

### Known issues with pull request:
* If review is out of bounds, the review top/bottom line scripts may not function as intended (i.e. they bring the cursor to the first/last visible range, not the first/last line of the buffer). This is likely needed to insure stability in the console.

### Change log entry:
== Changes ==
- Added a section to the user guide about NVDA in the Windows Console. (#9957)

### CODE REVIEW

**Comment by codeofdusk** [comment]
Cc @arfy8820 @feerrenrut @francipvb @leonardder

File: source/UIAUtils.py [hunk 1]
```diff
@@ -172,6 +172,24 @@ def getChildrenWithCacheFromUIATextRange(textRange,cacheRequest):
...
+		firstVisiRange = visiRanges.GetElement(0)
+		lastVisiRange = visiRanges.GetElement(visiLength - 1)
+		return range.CompareEndPoints(
+			UIAHandler.TextPatternRangeEndpoint_Start, firstVisiRange,
+			UIAHandler.TextPatternRangeEndpoint_Start
+		) < 0 or range.CompareEndPoints(
+			UIAHandler.TextPatternRangeEndpoint_Start, lastVisiRange,
+			UIAHandler.TextPatternRangeEndpoint_End) >= 0
+	else:
+		# Visible ranges not available, so fail gracefully.
```
Comment (leonardder) [inline source/UIAUtils.py]: Why not raise an error here and catch that when calling?

File: source/NVDAObjects/UIA/winConsoleUIA.py [hunk 2]
```diff
@@ -26,6 +27,21 @@ class consoleUIATextInfo(UIATextInfo):
 	#: to do much good either.
 	_expandCollapseBeforeReview = False
 
+	def __init__(self,obj,position,_rangeObj=None):
+		super(consoleUIATextInfo, self).__init__(obj, position, _rangeObj)
+		# Re-implement POSITION_FIRST and POSITION_LAST in terms of
```
Comment (leonardder) [inline source/NVDAObjects/UIA/winConsoleUIA.py]: Why should POSITION_First and POSITION_LAST be limited to text which is on screen? Wasn't the intention to also expose text that is off screen, or is that outside the scope of this pr now?

Comment (codeofdusk) [reply source/NVDAObjects/UIA/winConsoleUIA.py]: That is now out of scope.

File: source/UIAUtils.py [hunk 3]
```diff
@@ -172,6 +172,23 @@ def getChildrenWithCacheFromUIATextRange(textRange,cacheRequest):
 	c=CacheableUIAElementArray(c)
 	return c
 
+def isTextRangeOffscreen(range, visiRanges):
```
Comment (feerrenrut) [inline source/UIAUtils.py]: Please change the `range` argument so it does not shadow the builtin function

**Review by feerrenrut — CHANGES_REQUESTED** [review]
Overall the code looks ok, though from the PR description it's not really clear what use-cases this enables. After some playing with it I think I know what is going on, but it would be nice for others if there were steps (specifying key presses) to demonstrate.

While here I also think it is worth mentioning that you can also use incrementally select text with `shift+up/down/left/right arrows`, I was able to read many screens by: 
1. `shift+home`, to select back to the first column
1. `shift+up arrow`, to select previous lines. Of course the lines are in reverse order.
1. `shift+down arrow` to then unselect the text and hear it in the right order.

As we discussed selecting and jumping by word (`control+shift+left/right arrow`) is broken when NVDA is running.

Can you update the user guide to mention `control+up/down arrow` and `control+home/end` to scroll the console window, since it's not obvious how you would know about it.

Also as mentioned, please change the param name from `range` to `textRange`

**Comment by codeofdusk** [comment]
> Can you update the user guide to mention `control+up/down arrow` and `control+home/end` to scroll the console window, since it's not obvious how you would know about it.

Where in the user guide should this be mentioned? Also, since this is a Windows Console feature (not an NVDA one) should this still be in the NVDA user guide? (it seems that we tend not to mention application commands in the NVDA docs)

**Comment by feerrenrut** [comment]
> since this is a Windows Console feature (not an NVDA one) should this still be in the NVDA user guide?

It's true that we tend not to do this. However, it's quite a useful thing to know how to do when using NVDA, and not very discoverable. If there was not a built in feature, we would be trying to implement something similar I expect.

Overall I think a short guide to using the Windows Console is quite appropriate.

> Where in the user guide should this be mentioned?

You can add a subsection to `Application specific features`:

- Make sure that text that mentions "console" with context of windows console, links to this new subsection. 
- From the new subsection link back to `5.5. Reviewing Text`

**Comment by codeofdusk** [comment]
> > since this is a Windows Console feature (not an NVDA one) should this still be in the NVDA user guide?
> 
> It's true that we tend not to do this. However, it's quite a useful thing to know how to do when using NVDA, and not very discoverable. If there was not a built in feature, we would be trying to implement something similar I expect.
> 
> Overall I think a short guide to using the Windows Console is quite appropriate.
> 
> > Where in the user guide should this be mentioned?
> 
> You can add a subsection to `Application specific features`:
> 
> * Make sure that text that mentions "console" with context of windows console, links to this new subsection.
> * From the new subsection link back to `5.5. Reviewing Text`

Just updated. It's not exactly clear what control+home actually does though, so I didn't mention it yet.

**Comment by feerrenrut** [comment]
> It's not exactly clear what control+home actually does though, so I didn't mention it yet.

`control+home` scrolls the cmd window back to the first line in the (history/buffer).

File: user_docs/en/userGuide.t2t [hunk 4]
```diff
@@ -942,6 +942,19 @@ When in the table view of added books:
 | Context menu | applications | Opens the context menu for the selected book. |
 %kc:endInclude
 
+++ Windows Console ++[WinConsole]
+NVDA provides support for the Windows command console used by Command Prompt, PowerShell, and the Windows Subsystem for Linux.
+The console window is of fixed size: as new text is written, previous text is removed to make space.
```
Comment (feerrenrut) [inline user_docs/en/userGuide.t2t]:
```suggestion
The console window is of fixed size, typically much smaller than the buffer that holds the output.
As new text is written, the content scroll upwards and previous text is no longer visible. 
Text that is not visibly displayed in the window is not accessible with NVDA's text review commands.
```

**Review by feerrenrut — APPROVED** [review]

**Comment by XLTechie** [comment]
"new text is written, the content scroll upwards and previous text is no longer 
visible."

"Scroll" s/b "scrolls".

**Pull request merged**